pkg/client/unversioned/clientcmd/merged_client_builder.go

```diff
@@ -115,10 +115,10 @@ func (config *DeferredLoadingClientConfig) ClientConfig() (*restclient.Config, e
 		// TODO: this shouldn't be a global - the client config rules should be
 		//   handling this.
 		defaultConfig, defErr := DefaultClientConfig.ClientConfig()
-		if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
+		if defErr != nil && !IsConfigurationInvalid(defErr) {
 			return mergedConfig, nil
```

Comment:
> First, define what you mean by "non-host client" so I can make sure I'm following your scenario. 

A Kubernetes client (a binary built by using clientcmd libraries) that connects to an API server other than the API server of the pod that is running in.

> What inputs have been provided in your container?

Container is the `hyperkube` image. There is no special config provided. Kubeconfig for the client is supplied in a secret and is read through https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/client_config.go#L430

> The change I made for that was to the merge behavior for kubectl - if you can explain why your default config isnt invalid (everyone except kubectl should have empty default config) that may help identify the problem.

Our client isn't any special. It indeed has an empty default config as you describe here. But due to the the implementation detail in this code here, empty also means invalid. In other words, invalid error encompasses an empty error. So if you check for invalid, you are also inadvertently checking for emptiness. But from the way you are describing, you want one (validity check), but not the other (emptiness check). Specifically, a config is empty either when its `err == errConfigurationInvalid{ErrEmptyConfig}` or `err == ErrEmptyConfig`.

Even more specifically, your code here

```
if defErr != nil && !IsConfigurationInvalid(defErr) {
```

falls back to ICC when the default config has an error and is not invalid. But it also falls back to ICC when default config is empty, because `IsConfigurationInvalid(defErr)` says yes and the negation (`!IsConfigurationInvalid(defErr)`) says no. So you are off to other checks already, whereas you should have returned the `mergedConfig` when the default config said it was empty.

Also, the funny thing is this - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L1168 is merging `clientcmd.DefaultCluster` (which points to "http://localhost:8080") whereas the code here checks against `clientcmd.DefaultClientConfig` which is entirely different and has no Server field pointing to "http://localhost:8080".

By reading your description here, it feels like if you want to continue using `DefaultClientConfig` in these checks, you should also modify `DefaultClientConfig` here - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L1168 to include `DefaultCluster`. That way, you can drop this PR and it works in the non-host client cases because `DefaultClientConfig` is empty and/or invalid in that case and `mergedConfig` is returned. And in the case of kubectl, `DefaultClientConfig` which now has "https://localhost:8080" is both valid and non-empty, so we continue with the next check. The next check says go ahead and use the `mergedConfig` when it is not the same as default config. Otherwise it falls back to ICC.